kubeadm: Validate only the first cert entry when external ca mode is used

astundzia opened the pull request:
/sig cluster-lifecycle

#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
We are deploying kubernetes using kubeadm using an external ca cert file / external ca mode. This external ca cert file contains multiple certificates. When kubeadm runs in external CA mode, kubeadm runs validation logic that compares the entire existing ca certificate file (` /etc/kubernetes/pki/ca.crt`) to the newly rendered 'dummy' kubeconfig that only reads in the first certificate from `ca.crt`. The bytes are then compared & fails the check since they are different -- 1 contains 1+N certificates, the other only contains 1 certificate.

The `ca.crt` used within validation is read here:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/pkiutil/pki_helpers.go#L292

The original `ca.crt` is read here:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go#L257

The check is happening here:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go#L268

#### Which issue(s) this PR fixes:
Fixes kubernetes/kubeadm#3011

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```YES```

```release-note
kubeadm: during the validation of existing kubeconfig files on disk, handle cases where the "ca.crt" is a bundle and has intermediate certificates. Find a common trust anchor between the "ca.crt" bundle and the CA in the existing kubeconfig on disk instead of treating "ca.crt" as a file containing a single CA.
```

astundzia commented:
Working on adding the necessary EasyCLA auth with my org

astundzia commented:
Tested on kubedm 1.23.X and `master`.

astundzia commented:
EasyCLA completed

astundzia commented:
Added some test cases where a ca.crt contains a rootCA + intermediary CA.

neolit123 reviewed (commented):
thanks for the PR,
i will have a more detailed look next week.

i think such a change would need unit tests.
i will enable CI on the PR
/ok-to-test

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -264,8 +264,26 @@ func validateKubeConfig(outDir, filename string, config *clientcmdapi.Config) er
+	// Parse the expected certificate authority data
+	expectedCaCerts, err := certutil.ParseCertsPEM(caExpected)
+	if err != nil {
+		return errors.Errorf("the expected base64 encoded ca cert %q could not be parsed as a pem", caExpected)
```
```suggestion
		return errors.Errorf("the expected base64 encoded CA cert %q could not be parsed as a PEM", caExpected)
```

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -264,8 +264,26 @@ func validateKubeConfig(outDir, filename string, config *clientcmdapi.Config) er
+		return errors.Errorf("the expected base64 encoded ca cert %q could not be parsed as a pem", caExpected)
+	}
+
+	// only fetch the first certificate in the cacert. When this is read from file, only the first entry is considered
+	expectedCaCert := expectedCaCerts[0]
```
```suggestion
	// Only use the first certificate in the expected CA cert list
	expectedCACert := expectedCACerts[0]
```

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -264,8 +264,26 @@ func validateKubeConfig(outDir, filename string, config *clientcmdapi.Config) er
+	// only fetch the first certificate in the cacert. When this is read from file, only the first entry is considered
+	expectedCaCert := expectedCaCerts[0]
+
+	// Compare the current CA cert to the expected CA cert (which is only 1 entry).
```
```suggestion
	// Compare the current CA cert to the expected CA cert.
```

neolit123 reviewed (commented):
i think the change is fine.
added minor nits around wording and naming of vars.

please add a unit test.

neolit123 commented:
/triage accepted 
/priority backlog

astundzia commented:
Hey, thanks for the comments. I will address the comments & add a unit test this week. Thank you!

neolit123 commented:
/release-note-edit

```release-note
kubeadm: allow to validate "ca.crt" files that contain multiple certificates.
```

neolit123 commented:
> Hey, thanks for the comments. I will address the comments & add a unit test this week. Thank you!

@astundzia any updates?

astundzia commented:
Hey, working with the dev who ran into this issue at work to get the correct test cases. Please give me a few more days to get some tests up.

astundzia commented:
Added unit test, addressed comments. Please review when you get a chance. Thanks!

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig_test.go:
```diff
@@ -620,6 +618,20 @@ func TestValidateKubeconfigsForExternalCA(t *testing.T) {
 	// create a config with another server URL
 	configWithAnotherServerURL := setupdKubeConfigWithClientAuth(t, caCert, caKey, "https://4.3.2.1:4321", "test-cluster", "myOrg1")
 
+	// create a config with a ca cert containing multiple certificates
```
```suggestion
	// create a config with a CA cert containing multiple certificates
```

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig_test.go:
```diff
@@ -620,6 +618,20 @@ func TestValidateKubeconfigsForExternalCA(t *testing.T) {
 	configWithAnotherServerURL := setupdKubeConfigWithClientAuth(t, caCert, caKey, "https://4.3.2.1:4321", "test-cluster", "myOrg1")
 
+	// create a config with a ca cert containing multiple certificates
+	intermediateCaCert1, intermediateCaKey1 := certstestutil.SetupIntermediateCertificateAuthority(t, caCert, caKey)
+	intermediateCaCert2, intermediateCaKey2 := certstestutil.SetupIntermediateCertificateAuthority(t, intermediateCaCert1, intermediateCaKey1)
```
```suggestion
	intermediateCACert1, intermediateCAKey1 := certstestutil.SetupIntermediateCertificateAuthority(t, caCert, caKey)
	intermediateCACert2, intermediateCAKey2 := certstestutil.SetupIntermediateCertificateAuthority(t, intermediateCACert1, intermediateCAKey1)
```

to use the proper camelCase for abbreviations

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig_test.go:
```diff
@@ -620,6 +618,20 @@ func TestValidateKubeconfigsForExternalCA(t *testing.T) {
+	caCertBytes, _ := pkiutil.EncodeCertBundlePEM(caCertBundle)
+
+	clusterName := "myOrg1"
+	multipleCaConfig := setupdKubeConfigWithClientAuth(t, intermediateCaCert2, intermediateCaKey2, "https://1.2.3.4:1234", "test-cluster", clusterName)
```
```suggestion
	multipleCAConfig := setupdKubeConfigWithClientAuth(t, intermediateCaCert2, intermediateCaKey2, "https://1.2.3.4:1234", "test-cluster", clusterName)
```

neolit123 commented on cmd/kubeadm/app/util/certs/util.go:
```diff
@@ -43,6 +43,19 @@ func SetupCertificateAuthority(t *testing.T) (*x509.Certificate, crypto.Signer)
+// Intermediate CertificateAuthority cert/key pair
+func SetupIntermediateCertificateAuthority(t *testing.T, parentCert *x509.Certificate, parentKey crypto.Signer) (*x509.Certificate, crypto.Signer) {
+	caCert, caKey, err := pkiutil.NewIntermediateCertificateAuthority(parentCert, parentKey, &pkiutil.CertConfig{
+		Config: certutil.Config{CommonName: "kubernetes Intermediate CA"},
```
```suggestion
		Config: certutil.Config{CommonName: "kubernetes intermediate CA"},
```

neolit123 reviewed (commented):
added some more minor comments.
please keep the commit squashed to one.

a question that i have:

 i recall discussions that taking the first cert from a bundle maybe is arbitrary under certain conditions, we are already doing that in at least another location in the code base. but would it make more sense to compare entire bundles i.e. bundleCurrentCert1 == bundleExpCert1, undleCurrentCert2 == bundleExpCert12, ...

what do official specs say about that?

i don't think we should apply the change like that, but asking out of curiosity.
comparing only the first cert ensure we are comparing the CA root, supposedly, while intermediates follow and they can be perhaps different count between the current and expected.

astundzia commented:
Hey, I will address comments and squash the commits. I added some additional tests to cover this scenario.

The Root Certificate Authority (Root CA) is the fundamental element of the trust model. In terms of certificate validation, it's possible to create a Certificate Authority (CA) bundle that doesn't include the Root CA. As long as the CA bundle and the leaf certificates can trace back to a common trust anchor (a CA), the issued certificate should be valid.

However, you're correct that I should compare more than just the first certificate. Let's consider this (convoluted) scenario:

We have a Certificate Authority chain that includes the Root CA, Intermediate CA1, and Intermediate CA2. At some point, we issue Intermediate CA3 from Intermediate CA1. In this case, we can have a `ca.crt` that includes Intermediate CA1 and Intermediate CA2. The trust chain must include all certificates tracing back to a common trust anchor, which in this case is Intermediate CA1. Certificates issued from Intermediate CA3 should validate against a bundle containing any of the following combinations: {Root CA, Intermediate CA1, Intermediate CA2}, {Root CA, Intermediate CA1}, or {Intermediate CA1}, as long as the trust chain can be resolved.

In this scenario, it would be beneficial to compare all the current `ca.crt` certificates against the expected CA certificate. As long as there's a match, a trust chain should be established.

astundzia commented:
/retest

astundzia commented:
/retest

astundzia commented:
/retest

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig_test.go:
```diff
@@ -675,7 +677,138 @@ func TestValidateKubeconfigsForExternalCA(t *testing.T) {
+		},
+	}
+
+	// creates CA, write to pkiDir and remove ca.key to get into external CA condition
```
```suggestion
	// Creates CA, write to pkiDir and remove ca.key to get into external CA mode
```

that's how we call it (mode) across docs
also capitalization

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig_test.go:
```diff
@@ -675,7 +677,138 @@ func TestValidateKubeconfigsForExternalCA(t *testing.T) {
+	clusterName := "myOrg1"
+
+	// Create a kube config and assign the CA data to the CA bundle, issued from root CA
+	multipleCAConfigRootCAIssuer := setupdKubeConfigWithClientAuth(t, caCert, caKey, "https://1.2.3.4:1234", "test-cluster", clusterName)
```
a good opportunity to fix this typo?
```suggestion
	multipleCAConfigRootCAIssuer := setupKubeConfigWithClientAuth(t, caCert, caKey, "https://1.2.3.4:1234", "test-cluster", clusterName)
```

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig_test.go:
```diff
@@ -675,7 +677,138 @@ func TestValidateKubeconfigsForExternalCA(t *testing.T) {
+	// creates CA, write to pkiDir and remove ca.key to get into external CA condition
+	caCert, caKey := certstestutil.SetupCertificateAuthority(t)
+
+	// create a config with a CA cert containing multiple certificates
```
```suggestion
	// Create a config with a CA cert containing multiple certificates
```

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig_test.go:
```diff
@@ -675,7 +677,138 @@ func TestValidateKubeconfigsForExternalCA(t *testing.T) {
+	caCertBundleNoRootCA = append(caCertBundleNoRootCA, intermediateCACert1, intermediateCACert2)
+	caCertBytesNoRootCA, _ := pkiutil.EncodeCertBundlePEM(caCertBundleNoRootCA)
+
+	// Create a kube config and assign the CA data to the CA bundle, issued from intermediate CA 2,
```
```suggestion
	// Create a kubeconfig and assign the CA data to the CA bundle, issued from intermediate CA 2,
```

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig_test.go:
```diff
@@ -675,7 +677,138 @@ func TestValidateKubeconfigsForExternalCA(t *testing.T) {
+	multipleCAConfigIntermediateCA2Issuer := setupdKubeConfigWithClientAuth(t, intermediateCACert2, intermediateCAKey2, "https://1.2.3.4:1234", "test-cluster", clusterName)
+	multipleCAConfigIntermediateCA2Issuer.Clusters[clusterName].CertificateAuthorityData = caCertBytes
+
+	// create a config with a CA cert containing multiple certificates, omitting the root CA
```
```suggestion
	// Create a config with a CA cert containing multiple certificates, omitting the root CA
```

trying to use consistent capitalization of comments within a function (or even file)

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig_test.go:
```diff
@@ -675,7 +677,138 @@ func TestValidateKubeconfigsForExternalCA(t *testing.T) {
+	tests := map[string]struct {
+		filesToWrite  map[string]*clientcmdapi.Config
+		initConfig    *kubeadmapi.InitConfiguration
+		expectedError bool
```
all expected errors in this test are "false", i guess we should have at least one with "true"?

astundzia replied:
Added some negative tests. Reworked some additional logic. Will explain in more detail in a comment.

neolit123 commented on cmd/kubeadm/app/util/certs/util.go:
```diff
@@ -43,6 +43,19 @@ func SetupCertificateAuthority(t *testing.T) (*x509.Certificate, crypto.Signer)
+		Config: certutil.Config{CommonName: "kubernetes intermediate CA"},
+	})
+	if err != nil {
+		t.Fatalf("failure while generating intermediate CA certificate and key: %v", err)
```
```suggestion
		t.Fatalf("Failure while generating intermediate CA certificate and key: %v", err)
```

test output messages should be capitalized unlike non-test errors

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig_test.go:
```diff
@@ -675,7 +677,138 @@ func TestValidateKubeconfigsForExternalCA(t *testing.T) {
+	multipleCAConfigNoRootCA.Clusters[clusterName].CertificateAuthorityData = caCertBytesNoRootCA
+
+	if err := pkiutil.WriteCertBundle(pkiDir, kubeadmconstants.CACertAndKeyBaseName, caCertBundleNoRootCA); err != nil {
+		t.Fatalf("failure while saving CA certificate: %v", err)
```
```suggestion
		t.Fatalf("Failure while saving CA certificate: %v", err)
```

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -264,10 +264,35 @@ func validateKubeConfig(outDir, filename string, config *clientcmdapi.Config) er
+	// Parse the expected certificate authority data
+	expectedCACerts, err := certutil.ParseCertsPEM(caExpected)
+	if err != nil {
+		return errors.Errorf("the expected base64 encoded CA cert %q could not be parsed as a PEM", caExpected)
```
this would print the whole base64 string?
if so i think it should be at the end of the message:
```suggestion
		return errors.Errorf("the expected base64 encoded CA cert could not be parsed as a PEM:\n%q", caExpected)
```

generally users should not see this error as it would be a preparation error on our side - i.e. expected kubeconfig file was malformed. i think we already have tests for that.

neolit123 reviewed (commented):
thanks for the detailed explanation!
yes, i think that the extended validation against all current CAs makes sense.

the logic and new tests SGTM.
i did another pass over the code, found some minor prior nits and also some new ones.

i think i might request another reviewer to have a quick look, in case we missed something.

astundzia commented:
Hey, will get the comments addressed and a commit up in the next few days.

neolit123 commented:
> Hey, will get the comments addressed and a commit up in the next few days.

ok, note that code freeze for 1.30 is march 6th.

neolit123 commented:
> Hey, will get the comments addressed and a commit up in the next few days.

@astundzia do you have plans to finish this PR?

astundzia commented:
Hey, apologies for the lack of commits here. I will try to get this merged in this week!

astundzia commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -264,10 +264,35 @@ func validateKubeConfig(outDir, filename string, config *clientcmdapi.Config) er
+		}
 	}
+	if !trustAnchorFound {
+		return errors.Errorf("a kubeconfig file %q exists already but could not find trust anchor", kubeConfigFilePath)
```
Adjusted, `a kubeconfig file %q exists but does not contain a trusted CA in its current context's cluster. Total CA certificates found:`. I can reword if you think the trust anchor reference makes more sense, but I think most folks will see "trust anchor" and might be confused.

neolit123 replied:
ok, please fix the rest of the comments.
also keep the commits squashed to 1. the bots found a merge commit so the PR cannot merge as is
`do-not-merge/contains-merge-commits`

neolit123 commented:
error is:
```
{"Time":"2024-06-28T18:06:14.02662057Z","Action":"output","Package":"k8s.io/kubernetes/cmd/kubeadm/app/phases# k8s.io/kubernetes/cmd/kubeadm/app/phases/kubeconfig [k8s.io/kubernetes/cmd/kubeadm/app/phases/kubeconfig.test]
cmd/kubeadm/app/phases/kubeconfig/kubeconfig_test.go:761:34: undefined: setupdKubeConfigWithClientAuth
cmd/kubeadm/app/phases/kubeconfig/kubeconfig_test.go:765:43: undefined: setupdKubeConfigWithClientAuth
cmd/kubeadm/app/phases/kubeconfig/kubeconfig_test.go:769:43: undefined: setupdKubeConfigWithClientAuth
cmd/kubeadm/app/phases/kubeconfig/kubeconfig_test.go:779:30: undefined: setupdKubeConfigWithClientAuth
```
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/123102/pull-kubernetes-unit/1806748637615624192

astundzia commented:
Hey, I am working on this today, expect to have it up in an hour or so.

I redid a few things, and cleaned up some tests. Once I have it committed ill explain each test in diagrams for clarity.

astundzia committed:
When using external CA, look for common trust anchor within CA bundle.

astundzia commented:
Sorry for the delay. Please review when you can.

I've reworked the tests, and accurately read in all certificates within kubeconfig.go. I also addressed your comments, but a fair amount changed that I am sure I have some bad formatting again.

Logically, the tests are as follows. This is the certificate chain used in the tests at a high level.


<img width="401" alt="image" src="https://github.com/kubernetes/kubernetes/assets/831457/92022227-2a9d-40ce-839f-6d3795f49274">



For the first 3 tests, there is this CA certificate being used:
<img width="232" alt="image" src="https://github.com/kubernetes/kubernetes/assets/831457/6ce4571a-013a-49f3-bfef-254b7eba2437">

Test 1. Build a config using a leaf cert issued from the RootCA

Test 2. Build a config using a leaf cert issued from Intermediate CA 1A

Test 3. Build a config using a leaf cert issued from Intermediate CA 2A



For the next two tests, I build a CA cert that looks like this:

<img width="185" alt="image" src="https://github.com/kubernetes/kubernetes/assets/831457/6ce6c2ab-aede-434b-966d-78f21ecc9814">

Test 4. Build a config using a leaf cert issued from Intermediate CA 2B

Test 5. Build a config using a leaf cert issued from Intermediate CA 2A -- This should fail, it is issued from a different CA



For the next test, I wanted to make sure that something issued from a different CA still failed validation.

<img width="199" alt="image" src="https://github.com/kubernetes/kubernetes/assets/831457/2f7d98dc-498a-4ad6-ae8d-193ac5c43041">

Test 6. Build a config using a leaf cert issued from Intermediate CA 2B -- This should fail, it is issued from a different CA & chain is invalid

astundzia commented:
/retest

neolit123 commented:
/retest
flaky test problem is tracked here
https://github.com/kubernetes/kubernetes/pull/123102

neolit123 reviewed (commented):
thanks @astundzia 
i verified the changes a couple of times and this seems good to me.

/lgtm
/approve

/cc @SataQiu @carlory 
do you have any review comments on this?

neolit123 commented:
/uncc @chendave

neolit123 commented:
/release-note-edit

```release-note
kubeadm: during the validation of existing kubeconfig files on disk, handle cases where the "ca.crt" is a bundle and has intermediate certificates. Find a common trust anchor between the "ca.crt" bundle and the CA in the existing kubeconfig on disk instead of treating "ca.crt" as a file containing a single CA.
```

neolit123 commented:
> Sorry for the delay. Please review when you can.
> 
> I've reworked the tests, and accurately read in all certificates within kubeconfig.go. I also addressed your comments, but a fair amount changed that I am sure I have some bad formatting again.
> 
> Logically, the tests are as follows. This is the certificate chain used in the tests at a high level.

thanks for the detailed explanation, this helped me navigate the review!

neolit123 commented:
> /cc @SataQiu @carlory do you have any review comments on this?

forgot to put a hold. let's iterate in a new pr if needed.
